pkg/bindings/images/build.go

```diff
@@ -200,52 +228,116 @@ func Build(ctx context.Context, containerFiles []string, options entities.BuildO
 	go func() {
 		defer pw.Close()
+		defer gw.Close()
 		defer tw.Close()
 
 		for _, src := range sources {
-			s := src
-			err := filepath.Walk(s, func(path string, info os.FileInfo, err error) error {
+			s, err := filepath.Abs(src)
+			if err != nil {
+				logrus.Errorf("cannot stat one of source context: %v", err)
+				merr = multierror.Append(merr, err)
+				return
+			}
+
+			err = filepath.Walk(s, func(path string, info os.FileInfo, err error) error {
 				if err != nil {
 					return err
 				}
-				if !info.Mode().IsRegular() || path == s {
-					return nil
-				}
 
-				f, lerr := os.Open(path)
-				if lerr != nil {
-					return lerr
+				if path == s {
+					return nil // skip root dir
 				}
 
 				name := strings.TrimPrefix(path, s+string(filepath.Separator))
-				hdr, lerr := tar.FileInfoHeader(info, name)
-				if lerr != nil {
-					f.Close()
-					return lerr
+
+				excluded, err := pm.Matches(filepath.ToSlash(name)) // nolint:staticcheck
+				if err != nil {
+					return errors.Wrapf(err, "error checking if %q is excluded", name)
 				}
-				hdr.Name = name
-				if lerr := tw.WriteHeader(hdr); lerr != nil {
-					f.Close()
-					return lerr
+				if excluded {
+					return nil
 				}
 
-				_, cerr := io.Copy(tw, f)
-				f.Close()
-				return cerr
+				if info.Mode().IsRegular() { // add file item
+					f, lerr := os.Open(path)
+					if lerr != nil {
+						return lerr
+					}
+
+					hdr, lerr := tar.FileInfoHeader(info, name)
+					if lerr != nil {
+						f.Close()
+						return lerr
+					}
+					hdr.Name = name
+					if lerr := tw.WriteHeader(hdr); lerr != nil {
+						f.Close()
+						return lerr
+					}
+
+					_, cerr := io.Copy(tw, f)
+					f.Close()
+					return cerr
+				} else if info.Mode().IsDir() { // add folders
+					hdr, lerr := tar.FileInfoHeader(info, name)
+					if lerr != nil {
+						return lerr
+					}
+					hdr.Name = name
+					if lerr := tw.WriteHeader(hdr); lerr != nil {
+						return lerr
+					}
+				} else if info.Mode()&os.ModeSymlink != 0 { // add symlinks as it, not content
+					link, err := os.Readlink(path)
+					if err != nil {
+						return err
+					}
+					hdr, lerr := tar.FileInfoHeader(info, link)
+					if lerr != nil {
+						return lerr
+					}
+					hdr.Name = name
+					if lerr := tw.WriteHeader(hdr); lerr != nil {
+						return lerr
+					}
+				} //skip other than file,folder and symlinks
+				return nil
 			})
 			merr = multierror.Append(merr, err)
 		}
 	}()
 	return pr, merr
 }
+
+func parseDockerignore(root string) ([]string, error) {
+	ignore, err := ioutil.ReadFile(filepath.Join(root, ".dockerignore"))
+	if err != nil && !os.IsNotExist(err) {
+		return nil, errors.Wrapf(err, "error reading .dockerignore: '%s'", root)
+	}
+	rawexcludes := strings.Split(string(ignore), "\n")
+	excludes := make([]string, 0, len(rawexcludes))
+	for _, e := range rawexcludes {
+		if len(e) == 0 || e[0] == '#' {
+			continue
+		}
+		excludes = append(excludes, e)
+	}
+	return excludes, nil
+}
```

Comment:
Could/Should we call `./vendor/github.com/openshift/imagebuilder/builder.go:// ParseDockerIgnore ` instead?  Seems to be the same code, I'm just not sure if we should include an imagebuilder dependency, even though the Buildah code has one already that's in the build functionality.

Reply:
There is two reason for that. 

1) the --ignorefile parameter does not have effect. Firstly I tried to get IgnoreFile option parameter, however It is internal parameter of buildah. Normally this method should have a second paramter named as ignorefile. Then if the parameter is empty string it should use .dockerignore, else should use given ignore file. At my previous comments I mention that. podman build command uses buildah's bud commands parameters at its parameters however it does not handle all of them. Hence I previously asked how to get "--ignorefile" argument value. Then modify the code.

2) I don't know why, using imagebuilder.ParseDockerIgnore breaks down tests. especially vendoring. 

When I looked at buildah's bud command, the --ignorefile parsed at cmd and setted to the options.Excludes. We need to do same as that? I asked it before, However I didnot get a proper answer.

Reply:
./bin/podman-remote build --help | grep ignore
      --ignorefile string                            path to an alternate .dockerignore file

I guess the question is whether the ignorefile path should go over or the contents?

Reply:
No problem is, only buildah parses --ignorefile with https://github.com/containers/buildah/blob/59cd527d6d40b5a77d7f3e6560b90c2ef7102c93/cmd/buildah/bud.go#L357-L363 however podman doesnot do that. hence imagebuildah.BuildOptions.Excludes ara always empty. we should also put parsing ignorefile before https://github.com/containers/podman/blob/master/cmd/podman/images/build.go#L403

Reply:
Yes I think the parsing should happen on the client side.

Reply:
Ok I will add cmd and re-push pr.